### PR TITLE
Breaking Change: Cache Control For Blob Content

### PR DESCRIPTION
**Pull request by supermario-homely**
https://homelyau.atlassian.net/browse/HDW-5207

- Added optional cache control when adding blob content

### CODE REVIEW

File: src/Homely.Storage.Blobs/AzureBlob.cs [hunk 1]
```diff
@@ -171,21 +171,23 @@ await blob.ExistsAsync(cancellationToken))
         /// <inheritdoc />
         public async Task<string> AddAsync(object item,
                                            string blobId = null,
-                                           CancellationToken cancellationToken = default)
+                                           CancellationToken cancellationToken = default,
+                                           CacheControlType cacheControlType = CacheControlType.None)
```
Comment (PureKrome) [inline src/Homely.Storage.Blobs/AzureBlob.cs]: Put the cancellationToken at the end please. (yes, this could break)

and for the other methods, too.

**Review by PureKrome — CHANGES_REQUESTED** [review]
Also, need some tests, plz.

File: src/Homely.Storage.Blobs/AzureBlob.cs (hunk 1 above)
Comment (supermario-homely) [reply src/Homely.Storage.Blobs/AzureBlob.cs]: @PureKrome  that's done now :)

**Review by PureKrome — CHANGES_REQUESTED** [review]
Tests?

**Comment by supermario-homely** [comment]
@PureKrome  added tests. I didn't add one for the batch methods as they just call the `AddAsync` calls anyway.

File: src/Homely.Storage.Blobs/CacheControlType.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                case CacheControlType.NoCache:
+                    return "no-cache";
+                default:
+                    return string.Empty;
+            }
+        }
+
+        public static CacheControlType GetCacheControlType(string stringValue)
+        {
+            if (stringValue.Equals("no-store", StringComparison.OrdinalIgnoreCase))
```
Comment (PureKrome) [inline src/Homely.Storage.Blobs/CacheControlType.cs:36]: what happens if `stringValue` is null/empty? will this crash?

File: tests/Homely.Storage.Blobs.Tests/AddAsyncTests.cs [hunk 3]
```diff
@@ -140,5 +141,68 @@ public async Task GivenSomeUri_AddAsync_AddsTheUri()
...
+            // Act.
+            var blobId = await azureBlob.AddAsync(TestUser, cacheControlType: expectedCacheControl);
+
+            // Assert.
+            blobId.ShouldNotBeNullOrEmpty();
+            var blob = await azureBlob.GetAsync<SomeFakeUser>(blobId, new List<string> { "CacheControl" }, default);
+            blob.Data.ShouldLookLike(TestUser);
+            blob.MetaData.ShouldContainKey("CacheControl");
+
+            CacheControlType cacheControlResult = CacheControlTypeHelper.GetCacheControlType(blob.MetaData["CacheControl"].ToString());
```
Comment (PureKrome) [inline tests/Homely.Storage.Blobs.Tests/AddAsyncTests.cs]:
```suggestion
            var cacheControlResult = CacheControlTypeHelper.GetCacheControlType(blob.MetaData["CacheControl"].ToString());
```

File: tests/Homely.Storage.Blobs.Tests/AddAsyncTests.cs [hunk 4]
```diff
@@ -140,5 +141,68 @@ public async Task GivenSomeUri_AddAsync_AddsTheUri()
...
+        {
+            // Arrange.
+            var azureBlob = await GetAzureBlobAsync();
+
+            // Act.
+            var blobId = await azureBlob.AddAsync(TestUser, cacheControlType: expectedCacheControl);
+
+            // Assert.
+            blobId.ShouldNotBeNullOrEmpty();
+            var blob = await azureBlob.GetAsync<SomeFakeUser>(blobId, new List<string> { "CacheControl" }, default);
```
Comment (PureKrome) [inline tests/Homely.Storage.Blobs.Tests/AddAsyncTests.cs]: make the `new List<string> { "CacheControl" }` a private static variable, at the top of the file. then this (and other calls) can use the same instance.

File: tests/Homely.Storage.Blobs.Tests/AddAsyncTests.cs [hunk 5]
```diff
@@ -140,5 +141,68 @@ public async Task GivenSomeUri_AddAsync_AddsTheUri()
...
+            blob.Data.ShouldLookLike(TestUser);
+            blob.MetaData.ShouldContainKey("CacheControl");
+
+            blob.MetaData["CacheControl"].ShouldBeNull();
+        }
+
+        [Fact]
+        public async Task GivenAnObjectWhichHasUTF8CharactersButWeWantToAddItAsAscii_AddAsync_WithCacheControl_AddsTheObjectAsAsciiAndCacheControlSet()
```
Comment (PureKrome) [inline tests/Homely.Storage.Blobs.Tests/AddAsyncTests.cs]:
```suggestion
        public async Task GivenAnObjectWhichHasUTF8CharactersButWeWantToAddItAsAsciiAndWithCacheControl_AddAsync_AddsTheObjectAsAsciiAndCacheControlSet()
```

File: tests/Homely.Storage.Blobs.Tests/AddAsyncTests.cs [hunk 6]
```diff
@@ -140,5 +141,68 @@ public async Task GivenSomeUri_AddAsync_AddsTheUri()
                 memoryStream.Length.ShouldBeGreaterThan(0);
             }
         }
+
+        [Theory]
+        [InlineData(CacheControlType.NoCache)]
+        [InlineData(CacheControlType.NoStore)]
+        public async Task GivenAnObjectThatDoesNotExist_AddAsync_SetWithCache_AddsObjectAndCacheIsSet(CacheControlType expectedCacheControl)
```
Comment (PureKrome) [inline tests/Homely.Storage.Blobs.Tests/AddAsyncTests.cs]:
```suggestion
        public async Task GivenAnObjectThatDoesNotExistAndCacheControlIsSet_AddAsync__AddsTheObjectAndTheCacheIsSet(CacheControlType expectedCacheControl)
```

File: tests/Homely.Storage.Blobs.Tests/AddAsyncTests.cs [hunk 7]
```diff
@@ -140,5 +141,68 @@ public async Task GivenSomeUri_AddAsync_AddsTheUri()
...
+            blob.Data.ShouldLookLike(TestUser);
+            blob.MetaData.ShouldContainKey("CacheControl");
+
+            CacheControlType cacheControlResult = CacheControlTypeHelper.GetCacheControlType(blob.MetaData["CacheControl"].ToString());
+
+            cacheControlResult.ShouldBe(expectedCacheControl);
+        }
+
+        [Fact]
+        public async Task GivenAnObjectThatDoesNotExist_AddAsync_SetWithNoCache_AddsObjectAndCacheIsNotSet()
```
Comment (PureKrome) [inline tests/Homely.Storage.Blobs.Tests/AddAsyncTests.cs]:
```suggestion
        public async Task GivenAnObjectThatDoesNotExistAndCacheControlIsSet_AddAsync_AddsTheObjectAndTheCacheIsNotSet()
```

**Review by PureKrome — CHANGES_REQUESTED** [review]

File: src/Homely.Storage.Blobs/CacheControlType.cs (hunk 2 above)
Comment (supermario-homely) [reply src/Homely.Storage.Blobs/CacheControlType.cs:36]: good point! empty string won't crash but null will.. Fixed that up now.

File: tests/Homely.Storage.Blobs.Tests/AddAsyncTests.cs (hunk 4 above)
Comment (supermario-homely) [reply tests/Homely.Storage.Blobs.Tests/AddAsyncTests.cs]: 🔧

**Review by PureKrome — APPROVED** [review]

**Pull request merged**